src/vs/workbench/api/common/extHostNotebook.ts

```diff
@@ -969,3 +1005,143 @@ function createNotebookCellStatusBarApiItem(internalItem: NotebookCellStatusBarI
 		dispose() { internalItem.dispose(); }
 	});
 }
+
+enum NotebookCellExecutionTaskState {
+	Init,
+	Started,
+	Resolved
+}
+
+class NotebookCellExecutionTask extends Disposable {
+	private _onDidChangeState = new Emitter<void>();
+	readonly onDidChangeState = this._onDidChangeState.event;
+
+	private _state = NotebookCellExecutionTaskState.Init;
+	get state(): NotebookCellExecutionTaskState { return this._state; }
+
+	private readonly _tokenSource: CancellationTokenSource;
+	private _editsQueue = Promise.resolve<void | boolean>(undefined);
+
+	constructor(
+		private readonly _uri: vscode.Uri,
+		private readonly _index: number,
+		private readonly _document: ExtHostNotebookDocument,
+		private readonly _cell: ExtHostCell,
+		private readonly _proxy: MainThreadNotebookShape) {
+		super();
+		this._tokenSource = this._register(new CancellationTokenSource());
+
+		this.mixinMetadata({
+			runState: NotebookCellExecutionState.Pending,
+			lastRunDuration: undefined
+		});
+	}
+
+	cancel(): void {
+		this._tokenSource.cancel();
+	}
+
+	private async applyEdits(getEdits: () => ICellEditOperation[]): Promise<void> {
```

Comment:
This isn't correct as is. I could get rid of this queue if we could send a partial metadata object to the main thread as a change to apply, instead of a complete metadata object overwriting other properties but `ICellEditOperation` only takes a full metadata object. Should I just add a method to MainThreadNotebook that does exactly what I want or should I do something with ICellEditOperation? @rebornix?

Reply:
I would prefer ICellEditOperation.